content/blog/2017/2017-02-10-declarative-html-publisher.adoc

```diff
@@ -0,0 +1,322 @@
+    // build and run tests with coverage
+    sh 'bundle exec rake build spec'
+
+    // Archive the built artifacts
+    archive includes: 'pkg/*.gem'
+
+    // publish html
+    publishHTML [
+        allowMissing: false,
+        alwaysLinkToLastBuild: false,
+        keepAll: true,
+        reportDir: 'coverage',
+        reportFiles: 'index.html',
+        reportName: "RCov Report"
+      ]
+
+  }
+}
+----
+<1> Select where to run this Pipeline, in this case "any" agent, regardless of label.
+<2> Declarative automatically performs a checkout of source code on the agent,
+    whereas Scripted Pipeline users must explicitly call `checkout scm`.
+<3> Set the Pipeline option to preserve the ten most recent runs.
+This overrides the default behavior from the Multibranch parent of this Pipeline.
+<4> Run the "Build" stage.
+
+image::/images/post-images/2017-02-09/declarative-build.png[Stage view, role="center"]
+
+Now that we have this Pipeline in Declarative form, let's take a minute to do a
+little clean up.  We'll split out the `bundle` actions a little more and move
+steps into logically grouped stages.  Rather than having one monolithic "Build"
+stage, we'll have details for each stage.  As long as we're prettying things
+up, let's switch to using link:/projects/blueocean/[Blue Ocean] to view our
+runs, as well.
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+  agent any
+  options {
+    // Keep the 10 most recent builds
+    buildDiscarder(logRotator(numToKeepStr:'10'))
+  }
+  stages {
+    stage ('Install') {
+      steps {
+        // install required gems
+        sh 'bundle install'
+      }
+    }
+    stage ('Build') {
+      steps {
+        // build
+        sh 'bundle exec rake build'
+
+        // Archive the built artifacts
+        archive includes: 'pkg/*.gem'
+      }
+    }
+    stage ('Test') {
+      steps {
+        // run tests with coverage
+        sh 'bundle exec rake spec'
+
+        // publish html
+        publishHTML target: [
+            allowMissing: false,
+            alwaysLinkToLastBuild: false,
+            keepAll: true,
+            reportDir: 'coverage',
+            reportFiles: 'index.html',
+            reportName: "RCov Report"
+          ]
+      }
+    }
+  }
+}
+// Scripted //
+----
+
+image::/images/post-images/2017-02-09/declarative-stages.png[Blue Ocean View, role="center"]
+
+
+== Using `post` sections
+
+This looks pretty good, but if we think about it
+the `archive` and `publishHTML` steps are really post-stage actions.
+They should only occur when the rest of their stage succeeds.
+As our Pipeline gets more complex we might need to add actions that always happen
+even if a stage or the Pipeline as a whole fail.
+
+In Scripted Pipeline, we would use `try/catch/finally`,
+but we cannot do that in Declarative.
+One key feature of the Declarative Pipeline
+is that it does not allow script-based control structures
+such as `for` loops, `if-then-else` blocks, or `try-catch-finally` blocks.
```

Comment:
Minor: `try-catch-finally` should be `try/catch/finally`, to be consistent with a few lines earlier (or the other way around).